domain/resource/state/resource.go

```diff
@@ -407,198 +421,543 @@ WHERE uuid = $resourceIdentity.uuid`,
-	checkValidResourceStmt, err := st.Prepare(checkValidResourceQuery, resourceUUID)
+	kind, err := charmresource.ParseType(resKind.Name)
 	if err != nil {
-		return resource.SetUnitResourceResult{}, errors.Capture(err)
+		return 0, errors.Errorf("parsing resource kind: %w", err)
+	}
+	return kind, err
+}
+
+// insertFileResource checks that the storage ID corresponds to stored object
+// store metadata and then records that the resource is stored at the provided
+// storage ID.
+func (st *State) insertFileResource(
+	ctx context.Context,
+	tx *sqlair.TX,
+	resourceUUID coreresource.UUID,
+	storageID coreresourcestore.ID,
+) error {
+	// Get the object store UUID of the stored resource blob.
+	if !storageID.IsObjectStoreUUID() {
+		return errors.Errorf("cannot insert file resource that is not stored in the object store")
+	}
+	uuid, err := storageID.ObjectStoreUUID()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	// Check the resource blob is stored in the object store.
+	storedResource := storedFileResource{
+		ResourceUUID:    resourceUUID.String(),
+		ObjectStoreUUID: uuid.String(),
+	}
+	checkObjectStoreMetadataStmt, err := st.Prepare(`
+SELECT uuid AS &storedFileResource.store_uuid
+FROM object_store_metadata
+WHERE uuid = $storedFileResource.store_uuid
+`, storedResource)
+	if err != nil {
+		return errors.Capture(err)
 	}
 
-	// Prepare statement to verify if the application resource is already
-	// retrieved.
+	err = tx.Query(ctx, checkObjectStoreMetadataStmt, storedResource).Get(&storedResource)
+	if errors.Is(err, sqlair.ErrNoRows) {
+		return errors.Errorf("checking object store for resource %s: %w", resourceUUID, resourceerrors.StoredResourceNotFound)
+	} else if err != nil {
+		return errors.Errorf("checking object store for resource %s: %w", resourceUUID, err)
+	}
+
+	// Check if the resource has already been stored.
+	checkResourceFileStoreStmt, err := st.Prepare(`
+SELECT &storedFileResource.*
+FROM   resource_file_store
+WHERE  resource_uuid = $storedFileResource.resource_uuid
+`, storedResource)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	err = tx.Query(ctx, checkResourceFileStoreStmt, storedResource).Get(&storedResource)
+	if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+		return errors.Errorf("checking if resource %s already stored: %w", resourceUUID, err)
+	} else if err == nil {
+		// If a row was found, return that the resource is already stored.
+		return resourceerrors.ResourceAlreadyStored
+	}
+
+	// Record where the resource is stored.
+	insertStoredResourceStmt, err := st.Prepare(`
+INSERT INTO resource_file_store (*)
+VALUES ($storedFileResource.*)
+`, storedResource)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	var outcome sqlair.Outcome
+	err = tx.Query(ctx, insertStoredResourceStmt, storedResource).Get(&outcome)
+	if err != nil {
+		return errors.Errorf("resource %s: %w", resourceUUID, err)
+	}
+
+	if rows, err := outcome.Result().RowsAffected(); err != nil {
```

Comment:
**thought:** I still feel weird to see this kind of check, for me they are verbose. Counting affected rows in case of update make sense, however, but for an insert ? I think if the insert succeed without error, it is certain that you have one row inserted. And if not... That's probably due to a developer errors, which should eventually be catched by UT.

(And BTW, that's not seems to be consistently checked over the whole file.

Reply:
Yea, to be honest I agree with you. I think at some point I was advised to do it on all DML queries, but for stright inserts it really doesn't make much sense. If the insert is doing a nested select or something then it should definatly be done. I had missed one in this PR which I have now added, and removed the basic insert ones.